### PR TITLE
Autocompletions for fish shell

### LINKED ISSUE #74: Autocompletion in different shells
I'm a [`fish`](https://fishshell.com) user and I really love the autocompletion in it, so I'm also really happy that Marathon supports this for `zsh`.
I've decided that I'll extend Marathon's great user experience to `fish` and it's rather easy task, but some thoughts came to my mind during the investigation. Here they are...

---
First of all, `installShellAutocompleteIfNeeded ` installs `zsh` completions no matter what kind of the shell you're using, so `~/.marathon` looks like this (out of the box), even though I don't have `zsh` on my computer:
```
.marathon/
├── Packages
│   ├── Generated
│   └── Temp
├── Scripts
│   ├── Cache
│   └── Temp
└── zsh
    └── _marathon
```
I have 2 proposals:
1. `installShellAutocompleteIfNeeded` should check for the shell which user is actually using and install completions only for it.
2. Rename `.marathon/zsh` to `.marathon/autocompletions` to be more generic. 

I think that second option could be better since those completion files aren't big and they can't make any harm.

---
Another thing is that Marathon is using custom script from `ZshAutocompleteInstaller` to _generate_ a string which is later appended to the file. It's cool that we want to keep everything in Swift, but IMO, at some point this can become a huge pain. 
Both `fish` and `zsh` (and probably more) are using scripts to achieve autocompletion, so why not write them and keep them in repo, or as a dependency?
I'm also thinking about some sort of abstraction layer, so each completion can actually use one set of completions and then generate from this file a proper autocompletion scripts for each shell which Marathon will support. Just thinking in **DRY** mode right now 😅

### PR DESCRIPTION
**Pull request by cojoj**
This PR closes #74 

As discussed with @JohnSundell in #74 - it's super cool to have autocompletions in shells, so it's way easier to use tool. I've added _very basic_ for of autocompletions for `fish` shell.

zrzut ekranu 2017-05-31 o 19 56 22

During the development of this _feature_ I was hoping to find a universal solution, but in the end I think what we currently have is simply an overkill 😂 Let me tell you why...
(But there's a place for some nice refactoring, cause what I did is a code duplication. We can standardize it, extract it and make it way nicer)

---

The most popular shells are (I guess): `bash`, `zsh` and `fish`. In that case making something _super generic_ and fully automatic is just a something we shouldn't spend much time.

Secondly, shells like `fish` and `zsh` (AFAIR) are **interactive**, so you have a lot of nice features available there which we simply can't use, cause we're generating scripts, so we assume they're really repetitive, right?
Example: in `fish` you can specify behavior of the `completion` based on passed parameters and also specify the results via different command. In case of Marathon when we call completion on `marathon remove` it will suggest only installed packages at `.marathon/Scripts/Cached` - **COOL, HUH?** 😎

Last, but not least, when someone will want to contribute to this... Well... It's hard to get easily what's happening, cause we're escaping string and blah, blah, blah...

---

If you take a look at the file I had to write to generate `marathon.fish` and on `marathon.fish` directly...
```swift
internal final class FishAutocompleteInstaller {
    static func installIfNeeded(in folder: Folder) {
        do {
            guard !folder.containsSubfolder(named: "fish") else {
                return
            }

            let zshFolder = try folder.createSubfolder(named: "fish")
            let autocompleteFile = try zshFolder.createFile(named: "marathon.fish")

            var autocompleteCode = generateFishFunctions()

            for command in Command.all {
                autocompleteCode.append("complete -f -c marathon " +
                        "-n '__fish_marathon_needs_command' " +
                        "-a '\(command.rawValue)' " +
                        "-d '\(command.description)'" +
                        "\n")
            }

            try autocompleteFile.write(string: autocompleteCode)
        } catch {
            // Since this operation isn't critical, we silently fail if an error occur
        }
    }

    fileprivate static func generateFishFunctions() -> String {
        let needsCommand =  "function __fish_marathon_needs_command\n" +
                            "\tset cmd (commandline -opc)\n" +
                            "\tif [ (count $cmd) -eq 1 ]\n" +
                            "\t\treturn 0\n" +
                            "\tend\n" +
                            "\treturn 1\n" +
                            "end" +
                            "\n\n"

        return needsCommand
    }
}
```

```bash
function __fish_marathon_needs_command
	set cmd (commandline -opc)
	if [ (count $cmd) -eq 1 ]
		return 0
	end
	return 1
end

complete -f -c marathon -n '__fish_marathon_needs_command' -a 'create' -d 'Create new script at a given path and open it'
complete -f -c marathon -n '__fish_marathon_needs_command' -a 'edit' -d 'Edit a script at a given path'
complete -f -c marathon -n '__fish_marathon_needs_command' -a 'remove' -d 'Remove a package or the cache data for a script at a given path'
complete -f -c marathon -n '__fish_marathon_needs_command' -a 'run' -d 'Run a script at a given path'
complete -f -c marathon -n '__fish_marathon_needs_command' -a 'install' -d 'Install a script at a given path or URL as a binary'
complete -f -c marathon -n '__fish_marathon_needs_command' -a 'add' -d 'Add a package from a given URL to be able to use it from your scripts'
complete -f -c marathon -n '__fish_marathon_needs_command' -a 'list' -d 'List all packages and cached script data'
complete -f -c marathon -n '__fish_marathon_needs_command' -a 'update' -d 'Update all added packages to their latest versions'
complete -f -c marathon -n '__fish_marathon_needs_command' -a 'help' -d 'Print these instructions'
```

Simple math: **38 lines vs 17 lines**
I do realize that maintaining bash scripts isn't so cool compared to writing Swift scripts, but you get my point 😉 Also, this will require manual updates to each completions file just to make them work with the latest Marathon version, but still, I thinks it'll be a simpler solution 😉

Let me know what do you think... ✌️

### CODE REVIEW

File: Sources/MarathonCore/Marathon.swift [hunk 1]
```diff
@@ -40,8 +40,9 @@ public final class Marathon {
             let rootFolder = try fileSystem.createFolderIfNeeded(at: folderPath)
             let packageFolder = try rootFolder.createSubfolderIfNeeded(withName: "Packages")
             let scriptFolder = try rootFolder.createSubfolderIfNeeded(withName: "Scripts")
+            let autocompletionsFolder = try rootFolder.createSubfolderIfNeeded(withName: "Autocompletions")
```
Comment (JohnSundell) [inline Sources/MarathonCore/Marathon.swift]: How about naming this `ShellAutocomplete` instead? That way it's clear that it's for shells.

Comment (cojoj) [inline Sources/MarathonCore/Marathon.swift]: Makes sense 😉 We can do even more, but we need to discuss it more deeply 😉

Comment (JohnSundell) [inline Sources/MarathonCore/Marathon.swift]: Let's just do that for now in this PR; and then we can continue working on this in subsequent ones 😄 Just update that folder name and I'll merge this in 👍

**Comment by cojoj** [comment]
@JohnSundell I've changes the name (as discussed) and added unit test to check for existence 😉

**Pull request merged**